ee/desktop/menu/menu.go

```diff
@@ -1,36 +1,174 @@
 package menu
 
 import (
-	"fmt"
+	"os/exec"
+	"runtime"
 
 	"fyne.io/systray"
-	"github.com/kolide/kit/version"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
 	"github.com/kolide/launcher/ee/desktop/assets"
 )
 
-func Init(hostname string) {
-	onReady := func() {
-		systray.SetTemplateIcon(assets.KolideDesktopIcon, assets.KolideDesktopIcon)
-		systray.SetTooltip("Kolide")
+// MenuData encapsulates a systray icon and accessible menu items
+type MenuData struct {
+	Tooltip string         `json:"tooltip,omitempty"`
+	Items   []MenuItemData `json:"items"`
+}
 
-		versionItem := systray.AddMenuItem(fmt.Sprintf("Version %s", version.Version().Version), "")
-		versionItem.Disable()
+// MenuItemData represents a menu item, optionally containing sub menu items
+type MenuItemData struct {
+	Label       string         `json:"label,omitempty"`
+	Tooltip     string         `json:"tooltip,omitempty"`
+	Disabled    bool           `json:"disabled,omitempty"`
+	NonProdOnly bool           `json:"nonProdOnly,omitempty"`
+	IsSeparator bool           `json:"isSeparator,omitempty"`
+	Action      MenuItemAction `json:"action,omitempty"`
+	Items       []MenuItemData `json:"items,omitempty"`
+}
 
-		// if prod environment, return
-		if hostname == "k2device-preprod.kolide.com" || hostname == "k2device.kolide.com" {
-			return
-		}
+// MenuItemAction encapsulates what action should be performed when a menu item is invoked
+type MenuItemAction struct {
+	URL string `json:"url,omitempty"`
+}
+
+// MenuBuilder is an interface a menu parser can use to specify how the systray menu is built
+type MenuBuilder interface {
+	SetTooltip(tooltip string)
+	AddMenuItem(label, tooltip string, disabled, nonProdOnly bool, action MenuItemAction, parent any) any
+	AddSeparator()
+}
+
+type menu struct {
+	logger    log.Logger
+	hostname  string
+	doneChans []chan bool
+}
+
+func New(logger log.Logger, hostname string) *menu {
+	m := &menu{
+		logger:   logger,
+		hostname: hostname,
+	}
 
-		// in non prod environment
+	return m
+}
+
+func (m *menu) Init(buildMenu func()) {
+	// buildMenu will be invoked after the systray has been initialized
+	// Before the systray exits, cleanup the goroutines
+	systray.Run(buildMenu, m.cleanup)
+}
+
+func (m *menu) Build(menu *MenuData) {
+	// Remove all menu items each time we rebuild the menu
+	systray.ResetMenu()
```

Comment:
This might be a systray question, but does this cause the menu to blink through empty before being repopluated?

Reply:
If you have the menu engaged and visible it will quickly flicker. If the menu isn't visible, there's no visual change.